File: app/src/main/java/com/pr0gramm/app/feed/ContentType.java
package com.pr0gramm.app.feed;

import android.content.Context;
import android.os.Parcel;
import android.os.Parcelable;

import com.google.common.base.Joiner;
import com.google.common.base.Optional;
import com.google.common.collect.FluentIterable;
import com.pr0gramm.app.R;

import java.util.Collection;
import java.util.Set;

/**
 * Content type to load.
 */
public enum ContentType implements Parcelable {
    SFW(1, R.string.type_sfw), NSFW(2, R.string.type_nsfw), NSFL(4, R.string.type_nsfl);

    private final int flag;
    private final int title;

    ContentType(int flag, int desc) {
        this.flag = flag;
        this.title = desc;
    }

    public int getTitle() {
        return title;
    }

    public int getFlag() {
        return flag;
    }

    public static int combine(Iterable<ContentType> flags) {
        int sum = 0;
        for (ContentType flag : flags)
            sum += flag.getFlag();

        return sum;
    }

    /**
     * Gets a all the content types that are encoded in the given
     * flags number. This is the reverse of {@link #combine(Iterable)}.
     *
     * @param flags The encoded content types.
     */
    public static Set<ContentType> decompose(int flags) {
        return FluentIterable.of(values())
                .filter(ct -> (ct.flag & flags) != 0)
                .toSet();
    }

    /**
     * Returns the {@link com.pr0gramm.app.feed.ContentType} that matches the given
     * flag's value. There must be only one bit set on the flags parameter.
     * This returns an empty optional, if no content type could be found.
     */
    public static Optional<ContentType> valueOf(int flag) {
        return FluentIterable.of(values()).firstMatch(ct -> ct.flag == flag);
    }

    public static String toString(Context context, Collection<ContentType> types) {
        return FluentIterable.from(types)
                .transform(ContentType::getTitle)
                .transform(context::getString)
                .join(Joiner.on("+"));
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(ordinal());
    }

    public static final Parcelable.Creator<ContentType> CREATOR = new Parcelable.Creator<ContentType>() {
        public ContentType createFromParcel(Parcel source) {
            int idx = source.readInt();
            return ContentType.values()[idx];
        }

        public ContentType[] newArray(int size) {
            return new ContentType[size];
        }
    };
}

File: app/src/main/java/com/pr0gramm/app/feed/ContentType.kt
package com.pr0gramm.app.feed

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import com.pr0gramm.app.R
import com.pr0gramm.app.model.user.LoginState
import com.pr0gramm.app.parcel.creator
import java.util.EnumSet

/**
 * Content type to load.
 */
enum class ContentType constructor(val flag: Int, val title: Int) : Parcelable {
    SFW(1, R.string.type_sfw), NSFW(2, R.string.type_nsfw), NSFL(4, R.string.type_nsfl), NSFP(8, R.string.type_nsfp);

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, f: Int) {
        dest.writeInt(ordinal)
    }

    fun isAvailable(loginState: LoginState): Boolean {
        if (!loginState.authorized) {
            return this == SFW
        }

        if (!loginState.verified) {
            return this == SFW || this == NSFP
        }

        return true
    }

    companion object {
        val All: List<ContentType> = values().toList()
        val AllSet: EnumSet<ContentType> = EnumSet.allOf(ContentType::class.java)

        private val values: Array<ContentType> = values()

        fun combine(flags: Iterable<ContentType>): Int {
            return flags.sumOf { it.flag }
        }

        fun combine(vararg flags: ContentType): Int {
            return flags.sumOf { it.flag }
        }

        /**
         * Gets a all the content types that are encoded in the given
         * flags number. This is the reverse of [.combine].
         * @param flags The encoded content types.
         */
        fun decompose(flags: Int): Set<ContentType> {
            return All.filter { (it.flag and flags) != 0 }.toSet()
        }

        /**
         * Returns the first [ContentType] that is part of the given flags.
         */
        fun firstOf(flags: Int): ContentType {
            return All.firstOrNull { it.flag and flags != 0 } ?: SFW
        }

        /**
         * Returns the [com.pr0gramm.app.feed.ContentType] that matches the given
         * flag's value. There must be only one bit set on the flags parameter.
         * This returns an empty optional, if no content type could be found.
         */
        fun valueOf(flag: Int): ContentType? {
            return All.firstOrNull { it.flag == flag }
        }

        fun toString(context: Context, types: Collection<ContentType>): String {
            return types.map { context.getString(it.title) }.joinToString("+")
        }

        @JvmField
        val CREATOR = creator {
            val idx = it.readInt()
            values[idx]
        }
    }
}

/**
 * Removes implicitly added content types like [.NSFP]
 */
fun Set<ContentType>.withoutImplicit(): EnumSet<ContentType> {
    val types = EnumSet.copyOf(this)
    types.remove(ContentType.NSFP)
    return types
}
